ocaml/idl/ocaml_backend/gen_api.ml

```diff
@@ -126,6 +126,42 @@ let add_set_enums types =
           if List.exists (fun ty2 -> ty2 = DT.Set ty) types then [ty] else [DT.Set ty; ty]
         | _ -> [ty]) types)
 
+(* Returns a list of type sorted such that the first elements in the
+   list have nothing depending on them. Later elements in the list may
+   depend upon types earlier in the list *)
+let toposort_types highapi types =
+  let rec inner result remaining =
+    let rec references name = function
+      | DT.String
+      | DT.Int
+      | DT.Float
+      | DT.Bool
+      | DT.DateTime
+      | DT.Ref _
+      | DT.Enum _ -> false
+      | DT.Set ty -> references name ty
+      | DT.Map (ty, ty') -> (references name ty) || (references name ty')
+      | DT.Record record when record = name -> true
+      | DT.Record record ->
+        let all_fields = DU.fields_of_obj (Dm_api.get_obj_by_name highapi ~objname:record) in
+        List.exists (fun fld -> references name fld.DT.ty) all_fields
+    in
+    let (ty_ref,ty_not_ref) =
+      List.partition (fun ty -> match ty with
+      | DT.Record name ->
```

Comment:
Instead of
```OCaml
let referencing = List.filter (references name) remaining in
List.length referencing > 1
```
I think this could be
```OCaml
List.exists (fun t -> t <> ty && (references name t)) remaining
```
It would be quicker that way (not always checking every element in the `remaining` list) and (I feel) no harder to understand; possibly slightly easier.

Reply:
I quite like this, yes.